### ISSUE TITLE
NullReferenceException with FixSerializationForFullPersonName

### ISSUE THREAD

**Issue by hnguyen094** [lupidan/apple-signin-unity#110]
The purpose of `FixSerializationForObject` is to assign default of the type base on the bool param, and the default is `null` for classes.

I ran into an issue with that behavior here
https://github.com/lupidan/apple-signin-unity/blob/1bac913cf9c942c5a12c0a731d3a49ccf7c3a58b/AppleAuth/Native/AppleIDCredential.cs#L45-L47

Causing a `NullReferenceException` in
https://github.com/lupidan/apple-signin-unity/blob/1bac913cf9c942c5a12c0a731d3a49ccf7c3a58b/AppleAuth/Native/SerializationTools.cs#L25-L36

I ran into this bug on iPad Pro 2020, on iPadOS 15 Beta 2, where my payload json did have `"_hasFullName":0`. I could write a PR to add an additional null check in `FixSerializationForFullPersonName`, but wasn't sure if there are problems with other usages involving `FixSerializationForObject` returning `null`.

Anyway, thanks for the great plugin!

**Comment by lupidan** [comment]
So few things:

1. The class + bool pair used in "FixSerializationForObject" is used due to how the serialization system works in Unity, which does not deserialize considering null values (unless something was changed in recent Unity versions)
2. Technically, a null check would be required, but I am surprised to see this reported **so late**, you would think this was a more common issue, but it certainly looks like a bug. 

Is there any particular set of steps you do to reproduce it in your iOS Device? Like not requesting the users name, or have the name not be shared in the login dialog?

**Comment by hnguyen094** [comment]
Unfortunately I don't think there are any specific steps. I can tell you the specifics of what happened, in case that might help.

The setup for the app is:
1) Try Quick Login
2) If fails, do the full create account request with `fullName` field

I have 2 iPads: a 5th gen (ipad#1, iPadOS 14) and Pro 2nd gen (ipad#2, iPadOS 15 Beta 2 or 3). Both are using the same Apple ID.
iPad#1 both creates account/signs in perfectly fine, as expected.
iPad#2:
- when the app isn't signed in with Apple, I get
```
2021-07-15 11:51:34.764486-0400 CocoCast[10705:1511877] [Authorization] ASAuthorizationController credential request failed with error: Error Domain=com.apple.AuthenticationServices.AuthorizationError Code=1001 "(null)"
```
but the popup never appears and I never pressed cancel, so not sure how I get this error.
- when I use iPad#1 to sign the app in with Apple, iPad#2 now tries to `QuickLogin` successfully, but has the `NullReferenceException` error mentioned above after I press Continue (to acknowledge the sign in) on the popup. I haven't looked deep into the code to figure out why `QuickLogin` even checked for `fullName`, but hopefully this is somewhat helpful for you.

**Comment by lupidan** [comment]
Ok, this is an interesting one...

SO, for a normal Login, on an **iOS 14.4** device, if I don't request the user's name, Apple still returns an instance of `NSPersonNameComponents`, but that instance is pretty much empty, as in all is null...

```
Printing description of nameComponents:
<NSPersonNameComponents: 0x2831ffae0> {givenName = (null), familyName = (null), middleName = (null), namePrefix = (null), nameSuffix = (null), nickname = (null) phoneticRepresentation = (null) }
```

Therefore the sent payload from Objective-C to C# contains something like this:
´´´
{
    "_fullName" =     {
        "_hasPhoneticRepresentation" = 0;
    },
    "_hasFullName" = 1,
}
´´´

Basically, due to the way Apple is sending the data after the login request, we assume there is a name, even if it's empty.
Could it be that they have fixed this issue in newer iOS Beta versions?

In any case, I'll work on a proper fix for this issue ASAP. Expect a minor release soon.

Thanks for the report @hnguyen094

**Comment by lupidan** [comment]
@hnguyen094 A new release v1.4.2 with a fix for this bug is now available:

[Updated instructions](https://github.com/lupidan/apple-signin-unity#installation)
[Release page](https://github.com/lupidan/apple-signin-unity/releases/tag/v1.4.2)

Thanks again for the bug report. Let me know if this fixes the issue.

Closing this now.
